lib/src/_shared/PickerPopper.tsx

```diff
@@ -0,0 +1,143 @@
+import { useGlobalKeyDown, keycode } from './hooks/useKeyDown';
+import { IS_TOUCH_DEVICE_MEDIA } from '../constants/dimensions';
+import { executeInTheNextEventLoopTick } from '../_helpers/utils';
+import { TransitionProps } from '@material-ui/core/transitions/transition';
+
+export interface ExportedPickerPopperProps {
+  /**
+   * Popper props passed down to material-ui [Popper](https://material-ui.com/api/popper/#popper-api) component.
+   */
+  PopperProps?: Partial<PopperProps>;
+  /**
+   * Custom component for [Transition](https://material-ui.com/components/transitions/#transitioncomponent-prop).
+   */
+  TransitionComponent?: React.ComponentType<TransitionProps>;
+}
+
+export interface PickerPopperProps extends ExportedPickerPopperProps, PaperProps {
+  role: 'tooltip' | 'dialog';
+  TrapFocusProps?: Partial<TrapFocusProps>;
+  anchorEl: PopperProps['anchorEl'];
+  open: PopperProps['open'];
+  onClose: () => void;
+  onOpen: () => void;
+}
+
+export const useStyles = makeStyles(
+  theme => ({
+    popper: {
+      zIndex: theme.zIndex.modal,
+    },
+    paper: {
+      transformOrigin: 'top center',
+      '&:focus': {
+        outline: 'auto',
+        [IS_TOUCH_DEVICE_MEDIA]: {
+          outline: 0,
+        },
+      },
+    },
+    topTransition: {
+      transformOrigin: 'bottom center',
+    },
+  }),
+  { name: 'MuiPickersPopper' }
+);
+
+export const PickerPopper: React.FC<PickerPopperProps> = ({
+  role,
+  PopperProps,
+  TrapFocusProps,
+  TransitionComponent = Grow,
+  open,
+  innerRef = null,
+  anchorEl,
+  children,
+  onClose,
+  onOpen,
+  ...other
+}) => {
+  const classes = useStyles();
+  const lastFocusedElementRef = React.useRef<Element | null>(null);
+  const paperRef = React.useRef<HTMLElement>(null);
+  const handlePopperRef = useForkRef(paperRef, innerRef);
+
+  useGlobalKeyDown(open, {
+    [keycode.Esc]: onClose,
+  });
+
+  React.useEffect(() => {
+    if (role === 'tooltip') {
+      return;
+    }
+
+    if (open) {
+      lastFocusedElementRef.current = document.activeElement;
+    } else if (
+      lastFocusedElementRef.current &&
+      lastFocusedElementRef.current instanceof HTMLElement
+    ) {
+      lastFocusedElementRef.current.focus();
+    }
+  }, [open, role]);
+
+  const handleBlur = () => {
+    if (!open) {
+      return;
+    }
+
+    // document.activeElement is updating on the next tick after `blur` called
+    executeInTheNextEventLoopTick(() => {
+      if (paperRef.current?.contains(document.activeElement)) {
+        return;
+      }
+
+      onClose();
+    });
+  };
+
+  return (
+    <Popper
+      transition
+      role={role}
+      open={open}
+      anchorEl={anchorEl}
+      placement="bottom"
+      className={clsx(classes.popper, PopperProps?.className)}
+      {...PopperProps}
+    >
+      {({ TransitionProps, placement }) => (
+        <TrapFocus
+          open={open}
+          disableAutoFocus
+          disableEnforceFocus={role === 'tooltip'}
```

Comment:
Looking at the WAI-ARIA example, it seems we can disable enforce focus

```suggestion
          disableEnforceFocus
```

Reply:
It will make possible to tab out of <TrapFocus> which makes it completely useless

Reply:
By "tab out", are you referring to <kbd>Tab</kbd>? This sounds like an unrelated issue. The problem I see with the current logic is that if I click on another element of the page, it closes the popup but restores back the focus to calendar icon. Should we expect the focus to be set on the interacted element instead?

Reply:
I am playing locally with only `disableEnforceFocus` and it allows to tab out of picker and breaks TrapFocus at all

Reply:
Could it be an issue with the blur logic?

Reply:
I could reproduce the issue on https://codesandbox.io/s/lucid-davinci-om21n?file=/src/App.js. It does sounds like a bug with TrapFocus.

Reply:
https://github.com/mui-org/material-ui/issues/21569 in case you want to work on it 👌